jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecs/exec/EcsExecUtils.kt

```diff
@@ -95,6 +105,78 @@ object EcsExecUtils : CoroutineScope by ApplicationThreadPoolScope("EcsExec") {
         )
     }
 
+    fun getEc2InstanceTaskRoleArn(project: Project, clusterArn: String, ecsClient: EcsClient, task: software.amazon.awssdk.services.ecs.model.Task): String? {
+        try {
+            val iamClient = project.awsClient<IamClient>()
+            val containerInstanceArn = task.containerInstanceArn()
+            val res = ecsClient.describeContainerInstances(
+                DescribeContainerInstancesRequest
+                    .builder()
+                    .cluster(clusterArn)
+                    .containerInstances(containerInstanceArn).build()
+            )
+            val ec2InstanceId = res.containerInstances().first().ec2InstanceId()
+            val instanceProfileArn = project.awsClient<Ec2Client>().describeInstances(
+                DescribeInstancesRequest.builder().instanceIds(ec2InstanceId).build()
+            ).reservations().first().instances().first().iamInstanceProfile().arn() ?: return null
+            val instanceProfileName = instanceProfileArn.substringAfter(":instance-profile/")
+            return iamClient.getInstanceProfile(GetInstanceProfileRequest.builder().instanceProfileName(instanceProfileName).build()).instanceProfile()
+                .roles().first().arn() ?: null
+        } catch (e: Exception) {
+            return null
+        }
+    }
+
+    fun getTaskRoleArn(project: Project, clusterArn: String, taskArn: String): String? {
+        val ecsClient = project.awsClient<EcsClient>()
+        val task = ecsClient.describeTasks(DescribeTasksRequest.builder().tasks(taskArn).cluster(clusterArn).build()).tasks().first()
+        return if (task.overrides().taskRoleArn() != null) {
+            task.overrides().taskRoleArn()
+        } else {
+            val roleArn = project.getResourceNow(EcsResources.describeTaskDefinition(task.taskDefinitionArn())).taskRoleArn()
+            if (roleArn == null) {
+                val launchType = task.launchType()
+                if (launchType == LaunchType.EC2) {
+                    val ec2InstanceTaskRoleArn = getEc2InstanceTaskRoleArn(project, clusterArn, ecsClient, task)
+                    ec2InstanceTaskRoleArn
+                } else {
+                    return null
+                }
+            } else {
+                roleArn
+            }
+        }
```

Comment:
i think this would read better if we inverted the logic
```kt
val launchType = task.launchType()
when (launchType) {
    LaunchType.EC2 -> {
        // ec2 stuff
    }
    LaunchType.FARGATE -> {
        // fargate stuff
    }
    else -> {
        throw RuntimeException(message(/* launch type not supported */))
    }
}
```

Reply:
Since we only need to find out whether the launch type is EC2 to determine if it has a task role in its instance profile, and other launch types(fargate, unknown) need to be treated the same, an if else block can also work i think

Reply:
for the unknown case, we should probably throw instead of letting behavior fall through (for example, the [EXTERNAL launch type](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_Task.html#ECS-Type-Task-launchType) was released [recently](https://aws.amazon.com/blogs/aws/getting-started-with-amazon-ecs-anywhere-now-generally-available/)). the other thing im confused about is what happens if there is both an ec2 instance profile and task role arn defined in the task definition?

Reply:
In this instance, since we have to check the attributes and move forward only if the launch type is EC2, i was thinking of if-else . But you're right, I'll add separate cases for supported and unknown launch types.
According to the ECS Exec Checker script, (referred to by the ECS Exec team :https://github.com/aws-containers/amazon-ecs-exec-checker/blob/main/check-ecs-exec.sh#L467), this is the order for retrieving the task role:
1) Check if the task role is overriden in the task
2) if not, consider the task role in the task definition
3) If this is null and launch type is EC2, check for an instance profile and see if there's a task role there
Otherwise we display a warning saying that valid task role not found